src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -891,52 +891,113 @@ abstract class BTypes {
+    /**
+      * create a Lazy, whose calculation is performed with `frontendLock`
+      *
+      * @param t    the calculation
+      * @return a Lazy with which is guaranteed to perform if calculation with `frontendLock` locked
+      */
+    def withLock[T <: AnyRef](t: => T): Lazy[T] = new LazyWithLock[T](() => t)
+
+    /**
+      * create a Lazy where the result is pre-determined, typically a constant, e.g. Nil None etc
+      *
+      * @param value the final value of the Lazy
+      * @return a new Lazy, pre-initialised with the specified value
+      */
+    def eager[T <: AnyRef](value: T): Lazy[T] = new Eager[T](value)
 
-    private var initFunction = {
-      val tt = t // prevent allocating a field for t
-      () => { value = tt() }
+    /**
+      * create a Lazy, whose calculation is performed on demand
+      *
+      * @param t    the calculation
+      * @return a Lazy with which is guaranteed to perform if calculation with `frontendLock` locked
+      */
+    def withoutLock[T <: AnyRef](t: => T): Lazy[T] = new LazyWithoutLock[T](() => t)
+
+    private final class Eager[T](val force: T) extends Lazy[T] {
+      def onForce(f: T => Unit): Unit = f(force)
+
+      override def toString = force.toString
     }
 
-    override def toString = if (value == null) "<?>" else value.toString
+    private abstract class AbstractLazy[T <: AnyRef](t: () => T) extends Lazy[T] {
+      // value need be volatile to ensure that init doesn't expose incomplete results
+      // due to JVM inlining of init
+      @volatile protected var value: T = _
 
-    def onForce(f: T => Unit): Unit = {
-      if (value != null) f(value)
-      else frontendSynch {
+      override def toString = if (value == null) "<?>" else value.toString
+
+      private var postForce: List[T => Unit] = Nil
+
+      def onForce(f: T => Unit): Unit = {
         if (value != null) f(value)
+        else this.synchronized {
+          if (value != null) f(value)
+          else postForce ::= f
+        }
+      }
+
+      def force: T = {
+        // assign to local var to avoid volatile reads and associated memory barriers
+        var result = value
+        if (result != null) result
         else {
-          val prev = initFunction
-          initFunction = () => {
-            prev()
-            f(value)
+          result = init(t)
+          this.synchronized {
+            postForce foreach (_ (result))
+            postForce = Nil
           }
+          result
         }
       }
+
+      def init(t: () => T): T
     }
 
-    def force: T = {
-      if (value != null) value
-      else frontendSynch {
-        if (value == null) {
-          initFunction()
-          initFunction = null
-        }
+    /**
+      * A lazy value that synchronizes on the `frontendLock`, and supports accumulating actions
+      * to be executed when it's forced.
+      */
+    private final class LazyWithLock[T <: AnyRef](t: () => T) extends AbstractLazy[T](t) {
```

Comment:
In the previous implementation the init function (`t` here) was set to `null` after being used. Maybe we should keep that to release object references?

Reply:
done